src/wallet/walletdb.cpp

```diff
@@ -320,6 +320,72 @@ class CWalletScanState {
     CWalletScanState() = default;
 };
 
+bool LoadKey(CWallet* pwallet, DataStream& ssKey, DataStream& ssValue, std::string& strErr)
+{
+    LOCK(pwallet->cs_wallet);
+    try {
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/24914#discussion_r1235086722

> in [7be10ad](https://github.com/bitcoin/bitcoin/commit/7be10adff36c0dc49ae56ac571bb033cba7a565b): Can you explain why it is ok to put the try-catch here, where it previously wasn't? It looks like this will cause `ReadKeyValue` to not break out of the while-loop and instead continue?

IIUC, it is ok to catch this exception here because the same exception was also caught previously at the bottom of the ReadKeyValue function here:

https://github.com/bitcoin/bitcoin/blob/7be10adff36c0dc49ae56ac571bb033cba7a565b/src/wallet/walletdb.cpp#L781

and in both cases the exception handler just cause the ReadKeyValue function to just return false. There should be no case when an exception reading one key would cause ReadKeyValue function to throw and break out of the while loop in WalletBatch::LoadWallet. The [`catch (...)`](https://github.com/bitcoin/bitcoin/blob/7be10adff36c0dc49ae56ac571bb033cba7a565b/src/wallet/walletdb.cpp#L920) line around the while loop can catch exceptions, but not ones from `ReadKeyValue`

Reply:
> This has a bunch of consequences, for example a corrupt address book entry will now result in `Wallet corrupted` and the error string `All keys read correctly, but transaction data or address book entries might be missing or incorrect.` is now incorrect, or at least can no longer be hit with a corrupt address book entry.

I think it's preferable to treat all corruption errors as critical and fail to load the wallet. We can then loosen these over time. I think that would be better and easier than trying to reverse engineer when and what errors we would previously return.

Reply:
> I think it's preferable to treat all corruption errors as critical and fail to load the wallet. We can then loosen these over time.

This sounds good, but it would be good to say this explicitly in a code comment so it is clear what code is trying to do. It would also be good to update the error string if it is no longer accurate, and to describe in commit messages what behavior is changing, so we can know what changes are intended.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/24914#discussion_r1237398297

> This has a bunch of consequences, for example a corrupt address book entry will now result in `Wallet corrupted` [...]

Great catch! I didn't realize that when I was reviewing.

I think it actually would be preferable to just keep old behavior if that's easily possible, so it doesn't prevent loading wallets which previously just triggered warnings. But I can also see the case for being more strict. I mainly think if behavior is going to change, intent should clearly be documented in the code, and commit message should mention the change. Current commit eddfbcab498a8af200e006aef23fa43ad7c712cb "walletdb: refactor address book loading" description makes it seem like there is no change

Reply:
> Would it also make sense to add release notes to say that some wallets which previously had corrupt records won't be loaded by the new release? I could imagine panicking if a wallet that previously loaded with warnings no longer loaded, so release notes could clarify or say what to do. This is assuming affected wallets were usable before though..

Perhaps, but enumerating what failures result in which error messages is looking to be quite a non-trivial task for me. Looking at it more closely, I think that actually there's a lot of things that we would have said are non-critical errors but really should be critical, such as corrupted watchonly scripts and keys. These are things that would most definitely result in incorrect balances, but I think we would have allowed loading the wallet anyways. I think the wallet corruption error is a good enough message that people will open issues if they run into it, but I don't think we'll really see any of those since I think several of these should be causing actual other problems.

Reply:
> Perhaps, but enumerating what failures result in which error messages is looking to be quite a non-trivial task for me

That's understandable. I just meant a general release note like "Wallet database loading has changed in this new release, and wallets that could previously be loaded but had some corrupt records no may no longer load. If this happens it is recommended to load the wallet in the previous version of the software and import data into a new wallet, or report an issue." Idea would be to make it clear this is behavior we're aware of and not something to panic over.

Reply:
Added a release note.